### PR TITLE
Fixed #20347 -- Allowed customizing the maximum number of instantiated forms in formsets.

### PR DESCRIPTION
**Pull request by smithdc1**
[Ticket #20347](https://code.djangoproject.com/ticket/20347)

This follows on from #1593, I've updated the pull request based on the comments at the time and added release notes. 

Hopefully I've also added the original author as co-author correctly 🤞

### CODE REVIEW

File: docs/topics/forms/formsets.txt [hunk 1]
```diff
@@ -126,6 +126,46 @@ affect validation.  If ``validate_max=True`` is passed to the
...
+
+The ``absolute_max`` parameter to
+:func:`~django.forms.formsets.formset_factory` provides the ability to limit
+the maximum number of forms that can be instantiated when supplying ``POST``
+data.  This is to protect against memory exhaustion attacks using forged POST
+requests::
+
+    >>> from django.forms.formsets import formset_factory
+    >>> from myapp.forms import ArticleForm
+    >>> ArticleFormSet = formset_factory(ArticleForm, absolute_max=10)
```
Comment (felixxm) [inline docs/topics/forms/formsets.txt:144]: This raises ``ValueError('absolute_max must be..`` because `DEFAULT_MAX_NUM = 1000`.

File: tests/forms_tests/tests/test_formsets.py [hunk 2]
```diff
@@ -81,6 +81,13 @@ def prefixed(*args):
 
         return formset_class(data, **kwargs)
 
+    # Assert that the formset has the expected number of forms, and that each form within
+    # the formset has the proper prefix
+    def validate_formset_number_of_forms(self, formset, number_of_forms, prefix='form'):
+        self.assertEqual(number_of_forms, len(formset.forms))
+        for index in range(number_of_forms):
+            self.assertEquals('{0}-{1}'.format(prefix, index), formset.forms[index].prefix)
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:89]: I'm not sure if this assertions have value.

File: tests/forms_tests/tests/test_formsets.py [hunk 3]
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
...
+            'form-TOTAL_FORMS': '2505',
+            'form-INITIAL_FORMS': '0',
+            'form-MAX_NUM_FORMS': '0',
+        }
+
+        # The default absolute_max is max_num + DEFAULT_MAX_NUM (which is
+        # 2*DEFAULT_MAX_NUM if max_num isn't specified)
+        formset = FavoriteDrinksFormSet(data=data)
+        self.assertFalse(formset.is_valid())
+        self.validate_formset_number_of_forms(formset, 2 * formsets.DEFAULT_MAX_NUM)
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:901]: This test works without a patch, please move it to a separate commit/PR, we should also add assertion for `absolute_max`, e.g.
```python
    def test_default_absolute_max(self):
        # absolute_max defaults to 2 * DEFAULT_MAX_NUM if max_num is None.
        data = {
            'form-TOTAL_FORMS': 2001,
            'form-INITIAL_FORMS': '0',
            'form-MAX_NUM_FORMS': '0',
        }
        formset = FavoriteDrinksFormSet(data=data)
        self.assertIs(formset.is_valid(), False)
        self.assertEqual(
            formset.non_form_errors(),
            ['Please submit 1000 or fewer forms.'],
        )
        self.assertEqual(formset.absolute_max, 2000)
```

File: tests/forms_tests/tests/test_formsets.py [hunk 4]
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
...
+
+    def test_explicit_absolute_max_higher_than_default(self):
+        """
+        absolute_max can now be explicitly specified via kwarg to formset_factory.
+        This behavior was changed from absolute_max always being max_num + 1000 in the patch for #20347
+        """
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:907]: Chop this docstring.

File: tests/forms_tests/tests/test_formsets.py [hunk 5]
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
...
+        data = {
+            'form-TOTAL_FORMS': '2505',
+            'form-INITIAL_FORMS': '0',
+            'form-MAX_NUM_FORMS': '0',
+        }
+
+        # absolute_max can now be explicitly set, not being affected by
+        # max_num or DEFAULT_MAX_NUM at all
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:915]: I don't see any value in this comment.

File: tests/forms_tests/tests/test_formsets.py [hunk 6]
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
...
+        """
+        absolute_max can now be explicitly specified via kwarg to formset_factory.
+        This behavior was changed from absolute_max always being max_num + 1000 in the patch for #20347
+        """
+        data = {
+            'form-TOTAL_FORMS': '2505',
+            'form-INITIAL_FORMS': '0',
+            'form-MAX_NUM_FORMS': '0',
+        }
+
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:913]: Please chop unnecessary blank lines.

File: tests/forms_tests/tests/test_formsets.py [hunk 7]
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
...
+        # absolute_max can now be explicitly set, not being affected by
+        # max_num or DEFAULT_MAX_NUM at all
+        HigherLimitFavoriteDrinksFormSet = formset_factory(FavoriteDrinkForm, max_num=1, absolute_max=3000)
+        formset = HigherLimitFavoriteDrinksFormSet(data=data)
+        self.assertTrue(formset.is_valid())
+        self.validate_formset_number_of_forms(formset, 2505)
+
+        # verify that absolute_max still does provide a hard limit
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:921]: Please omit prefixes like "Verify that" since all tests verify things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

File: tests/forms_tests/tests/test_formsets.py [hunk 8]
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
...
+            'form-MAX_NUM_FORMS': '0',
+        }
+
+        LowerLimitFavoriteDrinksFormSet = formset_factory(FavoriteDrinkForm, max_num=30, absolute_max=30)
+        formset = LowerLimitFavoriteDrinksFormSet(data=data)
+        self.assertFalse(formset.is_valid())
+        self.validate_formset_number_of_forms(formset, 30)
+
+    def test_explicit_absolute_max_lower_than_default_no_max_num(self):
+        with self.assertRaises(ValueError):
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:941]: Use `assertRaisesMessage()`.

File: tests/forms_tests/tests/test_formsets.py [hunk 9]
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
...
+            formset_factory(FavoriteDrinkForm, absolute_max=30)
+
+    def test_max_num_greater_than_absolute_max(self):
+        with self.assertRaises(ValueError):
+            formset_factory(FavoriteDrinkForm, max_num=100, absolute_max=30)
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:946]: Use `assertRaisesMessage()`.

**Comment by smithdc1** [comment]
Hi @felixxm - Thanks for your feedback 🙏 

I've updated for your comments. I left the original commit alone and added my changes to subsequent commits. I was a bit unsure on rebasing for the release notes (as I had to force push?)

I've added some tests for ``modelformset_factory()`` and ``inlineformset_factory()`` but didn't think we need to test every permutation again?

File: docs/topics/forms/formsets.txt [hunk 10]
```diff
@@ -126,6 +126,38 @@ affect validation.  If ``validate_max=True`` is passed to the
 :func:`~django.forms.formsets.formset_factory`, then ``max_num`` will affect
 validation.  See :ref:`validate_max`.
 
+.. _formsets-absolute-max:
+
+Limiting the maximum number of instantiated forms
```
Comment (felixxm) [inline docs/topics/forms/formsets.txt:131]: I've changed `submitted` to `instantiated`.
